src/lang-promql/parser/matcher.ts

```diff
@@ -0,0 +1,87 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2020 The Prometheus Authors
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import { Subtree } from 'lezer-tree';
+import { EqlRegex, EqlSingle, LabelMatcher, LabelName, MatchOp, Neq, NeqRegex, StringLiteral } from 'lezer-promql';
+import { EditorState } from '@codemirror/next/basic-setup';
+
+class Matcher {
+  type: number;
+  name: string;
+  value: string;
+
+  constructor(type: number, name: string, value: string) {
+    this.type = type;
+    this.name = name;
+    this.value = value;
+  }
+
+  matchesEmpty(): boolean {
```

Comment:
You won't be able to detect matchers that match the empty string perfectly in Javascript, as you would need a JS-based RE2 (https://github.com/google/re2) implementation.

For example, `foo=~".*"` would also match empty labels although the regex string is not empty, and on the flip side, `foo!~"|bar"` does *not* match empty labels, although the regex string is not empty.

You could try matching the regexes in JS against an empty string (like we do in Go), but since JS regexes are different from RE2 regexes, this won't work perfectly either.

So maybe the best you can do is to only detect some cases, like `EqlSingle` and `Neq` ones, but not the regex cases (you could include some special-case hardcoded regex cases too, but it should never return a linter error for something that is not actually a matcher that matches an empty label).